src/realm/object-store/sectioned_results.cpp

```diff
@@ -0,0 +1,275 @@
+            current_section++;
+        }
+        else {
+            sections[section_key].indices.push_back(i);
+        }
+    }
+
+    std::vector<SectionRange> offset_ranges;
+    offset_ranges.reserve(sections.size());
+    for (auto& elem : sections) {
+        offset_ranges.push_back(std::move(elem.second));
+    }
+
+    auto desc = results.get_descriptor_ordering();
+    if (desc.will_apply_sort()) {
+        const SortDescriptor* sort_desc = static_cast<const SortDescriptor*>(desc[0]);
+        auto is_asc = sort_desc->is_ascending(0);
+        if (is_asc) {
+            bool is_ascending = *is_asc;
+            std::sort(offset_ranges.begin(), offset_ranges.end(), [&is_ascending](SectionRange a, SectionRange b) {
+                return is_ascending ? (a.key < b.key) : (a.key > b.key);
+            });
+        }
+    }
+    return offset_ranges;
+}
+
+static SectionedResults::SectionKeyFunc builtin_comparison(Results& results, Results::SectionedResultsOperator op,
+                                                           util::Optional<StringData> prop_name)
+{
+    switch (op) {
+        case Results::SectionedResultsOperator::FirstLetter:
+            if (results.get_type() == PropertyType::Object) {
+                return [p = *prop_name](Mixed value, SharedRealm realm) {
+                    auto obj = Object(realm, value.get_link());
+                    auto v = obj.get_column_value<StringData>(p);
+                    return v.size() > 0 ? v.prefix(1) : "";
+                };
+            }
+            else {
+                return [](Mixed value, SharedRealm) {
+                    auto v = value.get_string();
+                    return v.size() > 0 ? v.prefix(1) : "";
+                };
+            }
+        default:
+            throw std::logic_error("Builtin section algorithm not implemented.");
+    }
+}
+
+/// Returns the section of an element for the index of that element in the underlying Results.
+static SectionRange section_for_index(std::vector<SectionRange> offsets, size_t index)
+{
+    for (auto offset : offsets) {
+        if (std::find(offset.indices.begin(), offset.indices.end(), index) != offset.indices.end()) {
+            return offset;
+        }
+    }
+    throw std::logic_error("Section for given index not found.");
+}
+
+struct SectionedResultsNotificationHandler {
+public:
+    SectionedResultsNotificationHandler(SectionedResults& sectioned_results, SectionedResultsNotificatonCallback cb,
+                                        util::Optional<size_t> section_filter = util::none)
+        : m_cb(std::move(cb))
+        , m_sectioned_results(sectioned_results)
+        , m_prev_offset_ranges(m_sectioned_results.m_offset_ranges)
+        , m_section_filter(section_filter)
+    {
+    }
+
+    void before(CollectionChangeSet const&) {}
+    void after(CollectionChangeSet const& c)
+    {
+        m_sectioned_results.calculate_sections_if_required();
```

Comment:
This is something which ought to happen on the background thread. Sectioning the Results isn't a particularly cheap operation.